File: source/extensions/omni.isaac.sensor/python/nodes/OgnIsaacReadIMU.py
# Copyright (c) 2022-2023, NVIDIA CORPORATION. All rights reserved.
#
# NVIDIA CORPORATION and its licensors retain all intellectual property
# and proprietary rights in and to this software, related documentation
# and any modifications thereto. Any use, reproduction, disclosure or
# distribution of this software and related documentation without an express
# license agreement from NVIDIA CORPORATION is strictly prohibited.
#
# Array or tuple values are accessed as numpy arrays so you probably need this import
import numpy as np
import omni.graph.core as og
import omni.timeline
from omni.isaac.core_nodes import BaseResetNode
from omni.isaac.dynamic_control import _dynamic_control
from omni.isaac.sensor import _sensor
from omni.isaac.sensor.ogn.OgnIsaacReadIMUDatabase import OgnIsaacReadIMUDatabase


class OgnIsaacReadIMUInternalState(BaseResetNode):
    def __init__(self):
        self._is = _sensor.acquire_imu_sensor_interface()
        self.imu_path = ""
        super().__init__(initialize=False)

    def custom_reset(self):
        self.imu_path = ""
        pass


class OgnIsaacReadIMU:
    """
    Node that returns IMU Sensor data
    """

    @staticmethod
    def internal_state():
        return OgnIsaacReadIMUInternalState()

    @staticmethod
    def compute(db) -> bool:
        state = db.internal_state

        if len(db.inputs.imuPrim) > 0:
            state.imu_path = db.inputs.imuPrim[0].GetString()

        else:
            db.outputs.linAcc = [0.0, 0.0, 0.0]
            db.outputs.angVel = [0.0, 0.0, 0.0]
            db.outputs.orientation = [0.0, 0.0, 0.0, 1.0]
            db.outputs.sensorTime = 0.0
            db.log_error("Invalid Imu sensor prim")
            return False

        reading = state._is.get_sensor_reading(state.imu_path, None, db.inputs.useLatestData, db.inputs.readGravity)
        if reading.is_valid:
            db.outputs.orientation = reading.orientation
            lin_acc = [reading.lin_acc_x, reading.lin_acc_y, reading.lin_acc_z]
            ang_vel = [reading.ang_vel_x, reading.ang_vel_y, reading.ang_vel_z]
            db.outputs.sensorTime = reading.time
            db.outputs.linAcc = lin_acc
            db.outputs.angVel = ang_vel

        else:
            db.outputs.linAcc = [0.0, 0.0, 0.0]
            db.outputs.angVel = [0.0, 0.0, 0.0]
            db.outputs.orientation = [0.0, 0.0, 0.0, 1.0]
            db.outputs.sensorTime = 0.0
            db.log_warn("no valid sensor reading, is the sensor enabled?")
            return False
        db.outputs.execOut = og.ExecutionAttributeState.ENABLED

        return True

    @staticmethod
    def release(node):
        try:
            state = OgnIsaacReadIMUDatabase.per_node_internal_state(node)
        except Exception:
            state = None
            pass

        if state is not None:
            state.reset()

File: source/extensions/omni.isaac.sensor/nodes/OgnIsaacReadIMU.cpp
// Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
//
// NVIDIA CORPORATION and its licensors retain all intellectual property
// and proprietary rights in and to this software, related documentation
// and any modifications thereto. Any use, reproduction, disclosure or
// distribution of this software and related documentation without an express
// license agreement from NVIDIA CORPORATION is strictly prohibited.
//

// clang-format off
#include <UsdPCH.h>
// clang-format on

#include "omni/isaac/utils/UsdUtilities.h"

#include <omni/fabric/FabricUSD.h>
#include <pxr/base/gf/quatd.h>
#include <pxr/base/gf/vec3d.h>

#include <IsaacSensor.h>
#include <OgnIsaacReadIMUDatabase.h>


namespace omni
{
namespace isaac
{
namespace sensor
{

class OgnIsaacReadIMU
{
public:
    static void initialize(const GraphContextObj& contextObj, const NodeObj& nodeObj)
    {
        auto& state = OgnIsaacReadIMUDatabase::sInternalState<OgnIsaacReadIMU>(nodeObj);

        state.mImuSensorInterface = carb::getCachedInterface<ImuSensorInterface>();

        if (!state.mImuSensorInterface)
        {
            CARB_LOG_ERROR("Failed to acquire omni::isaac::sensor interface");
            return;
        }
    }

    static bool compute(OgnIsaacReadIMUDatabase& db)
    {
        auto& state = db.internalState<OgnIsaacReadIMU>();

        const auto& prim = db.inputs.imuPrim();
        const char* primPath;

        auto& linAcc = db.outputs.linAcc();
        auto& angVel = db.outputs.angVel();
        auto& orientation = db.outputs.orientation();
        auto& sensorTime = db.outputs.sensorTime();

        if (prim.size() > 0)
        {
            primPath = omni::fabric::toSdfPath(prim[0]).GetText();
        }
        else
        {
            linAcc = GfVec3d(0.0, 0.0, 0.0);
            angVel = GfVec3d(0.0, 0.0, 0.0);
            orientation = GfQuatd(1.0, 0.0, 0.0, 0.0);
            sensorTime = 0.0f;
            db.logError("Invalid Imu sensor prim");
            return false;
        }

        IsReading sensorReading = state.mImuSensorInterface->getSensorReading(
            primPath, nullptr, db.inputs.useLatestData(), db.inputs.readGravity());

        if (sensorReading.is_valid)
        {
            linAcc = GfVec3d(sensorReading.lin_acc_x, sensorReading.lin_acc_y, sensorReading.lin_acc_z);
            angVel = GfVec3d(sensorReading.ang_vel_x, sensorReading.ang_vel_y, sensorReading.ang_vel_z);
            orientation = GfQuatd(sensorReading.orientation.w, sensorReading.orientation.x, sensorReading.orientation.y,
                                  sensorReading.orientation.z);
            sensorTime = sensorReading.time;
        }
        else
        {
            linAcc = GfVec3d(0.0, 0.0, 0.0);
            angVel = GfVec3d(0.0, 0.0, 0.0);
            orientation = GfQuatd(1.0, 0.0, 0.0, 0.0);
            sensorTime = 0.0f;
            db.logWarning("no valid sensor reading, is the sensor enabled?");
            return false;
        }
        return true;
    }

private:
    ImuSensorInterface* mImuSensorInterface = nullptr;
};


REGISTER_OGN_NODE()
} // sensor
} // graph
} // omni
